torch_geometric/nn/models/lightgcn.py

```diff
@@ -0,0 +1,315 @@
+
+    Each layer's embeddings are computed as:
+
+    .. math::
+        \mathbf{x}^{l+1}_i = \sum_{j \in \mathcal{N}(i)}
+        \frac{1}{\sqrt{\deg(i)}\sqrt{\deg(j)}}\mathbf{x}^{l}_j
+
+    Two prediction heads are provided: :obj:`"ranking"` and
+    :obj:`"link_prediction"`. Ranking is performed by outputting the dot
+    product between the pair of embeddings chosen for evaluation. For link
+    prediction, the sigmoid of the ranking is computed to output label
+    probabilities.
+
+    .. note::
+
+        This implementation works not only on bipartite graphs but on any
+        kind of connected graph. Embeddings are propagated according to the
+        graph connectivity specified by :obj:`edge_index` while rankings or
+        probabilities are only computed (the forward pass) according to the
+        edges specified by :obj:`edge_label_index`.
+
+    Args:
+        num_layers (int): The number of Light Graph Convolution
+            (:obj:`LGConv`) layers.
+        num_nodes (int): The number of nodes in the graph.
+        embedding_dim (int, optional): The dimensionality of the node
+            embeddings (default: 32).
+        objective (str, optional): The objective of the model. The original
+            paper implements the :obj:`"ranking"` objective with a simple dot
+            product between embeddings as a prediction head. The option
+            :obj:`"link_prediction"` passes the ranking result through a
+            sigmoid layer (:obj:`"ranking"`, :obj:`"link_prediction"`)
+            (default: :obj:`"ranking"`).
+        lambda_reg (int, optional): The :math:`L_2` regularization strength of
+            the Bayesian Personalized Ranking (BPR) loss for the "ranking"
+            objective (default: 1e-4).
+        alpha (float, optional): The vector specifying the layer weights in the
+            final embeddings. If set to :obj:`None`, the original paper's
+            uniform initialization of :obj:`(1 / num_layers + 1)` is used
+            (default: :obj:`None`).
+        **kwargs (optional): Additional arguments of the underlying
+            :class:`torch_geometric.nn.conv.MessagePassing` layers.
+    """
+
+    objectives = ['ranking', 'link_prediction']
+
+    def __init__(self, num_layers: int, num_nodes: int,
+                 embedding_dim: int = 32, objective: str = 'ranking',
+                 lambda_reg: float = 1e-4, alpha: Tensor = None, **kwargs):
+        super().__init__(**kwargs)
+
+        assert (num_layers >= 1), 'Number of layers must be larger than 0.'
+        assert (num_nodes >= 1), 'Number of nodes must be larger than 0.'
+        assert (embedding_dim >= 1), 'Embedding dimension must be larger ' \
+            'than 0.'
+        assert (objective in LightGCN.objectives), 'Objective must be one ' \
+            f'of {LightGCN.objectives}'
+
+        if alpha is not None:
+            assert (alpha.size(0) == (num_layers + 1)), 'alpha must be of ' \
```

Comment:
We need to wrap `alpha` into a `Buffer` so that it gets correctly moved to GPU upon `model.cuda()`.